src/lib/expression/evaluation/expression_evaluator.cpp

```diff
@@ -0,0 +1,1054 @@
+    }
+
+    case PredicateCondition::In:
+      return _evaluate_in_expression<int32_t>(static_cast<const InExpression&>(predicate_expression));
+
+    case PredicateCondition::Like:
+    case PredicateCondition::NotLike:
+      return _evaluate_like_expression<int32_t>(static_cast<const BinaryPredicateExpression&>(predicate_expression));
+
+    case PredicateCondition::IsNull:
+    case PredicateCondition::IsNotNull:
+      return _evaluate_is_null_expression<int32_t>(static_cast<const IsNullExpression&>(predicate_expression));
+  }
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_predicate_expression(
+    const AbstractPredicateExpression& expression) {
+  Fail("Can only evaluate predicates to int32_t (aka bool)");
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_column_expression(
+    const PQPColumnExpression& column_expression) {
+  Assert(_chunk, "Cannot access Columns in this Expression as it doesn't operate on a Table/Chunk");
+
+  const auto& column = *_chunk->get_column(column_expression.column_id);
+  Assert(column.data_type() == data_type_from_type<R>(), "Can't evaluate column to different type");
+
+  _ensure_column_materialization(column_expression.column_id);
+  return std::static_pointer_cast<ExpressionResult<R>>(_column_materializations[column_expression.column_id]);
+}
+
+template <typename R>
+std::shared_ptr<ExpressionResult<R>> ExpressionEvaluator::_evaluate_case_expression(
+    const CaseExpression& case_expression) {
+  const auto when = evaluate_expression_to_result<int32_t>(*case_expression.when());
+
+  std::shared_ptr<ExpressionResult<R>> result;
+
+  _resolve_to_expression_results(
+      *case_expression.then(), *case_expression.else_(), [&](const auto& then_result, const auto& else_result) {
+        using ThenResultType = typename std::decay_t<decltype(then_result)>::Type;
+        using ElseResultType = typename std::decay_t<decltype(else_result)>::Type;
+
+        const auto result_size = _result_size(when->size(), then_result.size(), else_result.size());
+        std::vector<R> values(result_size);
+        std::vector<bool> nulls(result_size);
+
+        // clang-format off
+        if constexpr (Case::template supports<R, ThenResultType, ElseResultType>::value) {
+          for (auto chunk_offset = ChunkOffset{0};
+               chunk_offset < result_size; ++chunk_offset) {
+            if (when->value(chunk_offset) && !when->null(chunk_offset)) {
+              values[chunk_offset] = to_value<R>(then_result.value(chunk_offset));
+              nulls[chunk_offset] = then_result.null(chunk_offset);
+            } else {
+              values[chunk_offset] = to_value<R>(else_result.value(chunk_offset));
+              nulls[chunk_offset] = else_result.null(chunk_offset);
+            }
+          }
+        } else {
+          Fail("Illegal operands for CaseExpression");
```

Comment:
Guard clause?

Reply:
Not possible, sadly. `static_assert` would always trigger and an early return would still give us compiler errors, because illegal code would be generated.